javalib/src/main/scala/java/lang/String.scala

```diff
@@ -1,42 +1,890 @@
 package java.lang
 
-import java.io.Serializable
-
-class _String private() extends CharSequence with Serializable with Comparable[_String] {
-  def this(value: Array[scala.Char], offset: scala.Int, count: scala.Int) = this()
-
-  def length(): scala.Int = ???
-  def charAt(index: scala.Int): Char = ???
-  def compareTo(another_String: _String): scala.Int = ???
-  def subSequence(beginIndex: scala.Int, endIndex: scala.Int): CharSequence = ???
-  def concat(str: _String): _String = ???
-  def indexOf(ch: scala.Int): scala.Int = ???
-  def indexOf(ch: scala.Int, fromIndex: scala.Int): scala.Int = ???
-  def lastIndexOf(ch: scala.Int): scala.Int = ???
-  def lastIndexOf(ch: scala.Int, fromIndex: scala.Int): scala.Int = ???
-  def substring(beginIndex: scala.Int): _String = ???
-  def substring(beginIndex: scala.Int, endIndex: scala.Int): _String = ???
-  def startsWith(prefix: _String): scala.Boolean = ???
-  def toLowerCase(): _String = ???
-  def toCharArray(): Array[Char] = ???
-  def equalsIgnoreCase(str: _String): scala.Boolean = ???
-  def getChars(srcBegin: scala.Int, srcEnd: scala.Int, dst: Array[scala.Char],
-      dstBegin: scala.Int): Unit = ???
-  override def hashCode(): scala.Int = ???
-
-  // TODO: rest of the api
+import java.io.{Serializable, UnsupportedEncodingException}
+import java.util._
+import java.util.regex._
+import java.nio._
+import java.nio.charset._
+
+final class _String private ()
+    extends Serializable with Comparable[_String] with CharSequence {
+  private var value         = new Array[Char](0)
+  private var offset        = 0
+  private var count         = 0
+  private var cashedHashCode: Int = _
+
+  def this(data: Array[scala.Byte], _high: Int, start: Int, length: Int) = {
+    this()
+    if (length <= data.length - start && start >= 0 && 0 <= length) {
+      offset = 0
+      value = new Array[Char](length)
+      var high = _high
+      high <<= 8
+
+      var i = 0
+      while (i < length) {
+        value(i) = (high + data(start + i) & 0xff).toChar
+        i += 1
+      }
+    } else {
+      throw new StringIndexOutOfBoundsException()
+    }
+  }
+
+  def this(
+      data: Array[scala.Byte], start: Int, length: Int, encoding: Charset) = {
+    this()
+    if (encoding == null) {
+      throw new NullPointerException()
+    } else {
+      offset = 0
+      _String.lastCharset = encoding
```

Comment:
I'm skeptical about this `lastCharset` business. It looks like premature optimization to me, and it's not even clear that it's an optimization. An application sensibly working always with `Charset`s (rather than charset names) would just pay the cost of storing this global field every time for nothing.

Besides, if there's anywhere this caching should be done, it should be in `Charset.forName`, not in every single method that takes a charset name as parameter.